idseq_dag/steps/generate_coverage_viz.py

```diff
@@ -0,0 +1,241 @@
+
+        # Get a map from accession name to hits (reads and contigs)
+        # Also get a map from taxons to accession names.
+        (accession_data, taxons_to_accessions) = self.generate_accession_data(hit_summary, valid_contigs_with_read_counts)
+        # Remove taxons with no contigs in any of their accessions.
+        coverage_utils.remove_taxons_with_no_contigs(accession_data, taxons_to_accessions)
+
+        # Add the total_length and name of the accession to the accession data.
+        self.augment_accession_data_with_info(info_dict, accession_data)
+
+        # Get unassigned reads. Use a set for performance.
+        unassigned_reads_set = coverage_utils.get_unassigned_reads_set(accession_data)
+
+        # Extract information about contigs and reads.
+        contigs_map = self.generate_contigs_map(blast_top_m8, valid_contigs_with_read_counts)
+        reads_map = self.generate_reads_map(gsnap_deduped_m8, unassigned_reads_set)
+
+        # Add coverage to the contig data.
+        self.augment_contigs_map_with_coverage(contig_coverage_json, contigs_map)
+
+        # Add byteranges to the contig data.
+        self.augment_contigs_map_with_byteranges(contigs_fasta, contigs_map)
+
+        # Before selecting the best accessions and removing the rest, get the total accession count for each taxon.
+        taxons_to_total_accession_count = coverage_utils.get_taxons_to_total_accession_count(taxons_to_accessions)
+
+        # Select the best accessions for each taxon.
+        (taxons_to_accessions, accession_data) = coverage_utils.select_best_accessions_per_taxon(
+            taxons_to_accessions, accession_data, contigs_map, reads_map, num_accessions_per_taxon
+        )
+
+        # For each accession, generate the JSON that will be sent to the coverage_viz.
+        coverage_viz_obj = coverage_utils.generate_coverage_viz_json(accession_data, contigs_map, reads_map, max_num_bins_coverage)
+
+        # Generate the summary JSON which will get to the report page. JSON contains a map of taxons to valid accessions.
+        coverage_viz_summary_json = coverage_utils.generate_coverage_viz_summary_json(taxons_to_accessions, accession_data, coverage_viz_obj, taxons_to_total_accession_count)
+
+        # Write the summary JSON.
+        with open(coverage_viz_summary, 'w') as cvs:
+            json.dump(coverage_viz_summary_json, cvs)
+
+        # Create a coverage viz JSON file for each accession
+        coverage_viz_dir = os.path.join(self.output_dir_local, "coverage_viz")
+        command.execute(f"mkdir -p {coverage_viz_dir}")
+        for accession_id in coverage_viz_obj:
+            upload_file = os.path.join(coverage_viz_dir, f"{accession_id}_coverage_viz.json")
+
+            with open(upload_file, 'w') as uf:
+                json.dump(coverage_viz_obj[accession_id], uf)
+
+            self.additional_files_to_upload.append(upload_file)
```

Comment:
This needs to be changed to use `additional_folders_to_upload` instead of `additional_files_to_upload`. The former is a mechanism I just introduced in my last PR. It uploads everything in the folder to S3 in a single AWS CLI request. The latter uploads every file using an individual request. For thousands and thousands of files it's going to take hours.

Reply:
Please review the optimizations I introduced to cut the experimental stage from 3 days to 3 hours here: https://github.com/chanzuckerberg/idseq-dag/pull/139